### PR TITLE
[WOR-456] run sonarqube scan in build-and-test

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [WOR-456](https://broadworkbench.atlassian.net/browse/WOR-456)
* still need to add the `SONAR_TOKEN` secret to the repo
* follows [Appsec documentation](https://app.gitbook.com/o/-LVj1clrBdFW94pG2z2w/s/-LsrggqZs3CYGFRhFZdX-887967055/appsec-team-internal/appsec-team-internal/security-activities/sast-1) and configuration found in ECM and DrsHub

### CODE REVIEW

**Comment by TomConner** [comment]
Secret is coming soon, just needs some Atlantis

**Comment by TomConner** [comment]
Optional: add a badge to README.md…  `[Quality Gate Status](https://sonarcloud.io/summary/new_code?id=DataBiosphere_terra-billing-profile-manager)`

**Comment by TomConner** [comment]
Standby for https://github.com/broadinstitute/terraform-ap-deployments/pull/759

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -27,9 +27,22 @@ jobs:
...
+        with:
+          path: ~/.sonar/cache
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
+
       - name: Gradle build service
         run: ./gradlew --build-cache :service:build -x test
 
+      - name: SonarQube scan
+        run: ./gradlew --build-cache :service:sonarqube
```
Comment (pshapiro4broad) [inline .github/workflows/build-and-test.yml:41]: Running sonarqube this way won't pick up coverage information from your build. If you want coverage data as part of the sonar scan, you must run it after running `test`.

The good news is, once you've done this you can remove `codecov` from your workflow.

For an example, see https://github.com/DataBiosphere/terra-data-catalog/blob/main/.github/workflows/build-and-test.yml#L121

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -27,9 +27,22 @@ jobs:
           key: v1-${{ runner.os }}-gradle-${{ github.ref }}-${{ github.sha }}
           restore-keys: v1-${{ runner.os }}-gradle-${{ github.ref }}
 
+      - name: Cache SonarCloud packages
+        uses: actions/cache@v2
```
Comment (pshapiro4broad) [inline .github/workflows/build-and-test.yml:31]: Is there a performance benefit for using this cache action? We don't use `actions/cache` in our workflows and I'm wondering if we should start doing it. The sonar scan in our build takes less than 1min so it's already pretty fast.

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (pshapiro4broad) [reply .github/workflows/build-and-test.yml:41]: It looks like you also need to configure XML coverage output from `jacoco` if you aren't doing that already.

https://github.com/DataBiosphere/terra-data-catalog/blob/main/buildSrc/src/main/groovy/bio.terra.catalog.java-common-conventions.gradle#L97

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (TomConner) [reply .github/workflows/build-and-test.yml:31]: Whatever works. This came in with the template action provided by SonarCloud but you're welcome to change it. I don't see why there would be a benefit to caching actually if it's running one scan on an action runner that's going to be thrown away.

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (TomConner) [reply .github/workflows/build-and-test.yml:41]: Good ideas! Perhaps I can update the recommendations. From AppSec's point of view, we don't require coverage reporting but it is nice in the sonarqube status, and if it is better than codecov then hey why not.

**Review by TomConner — APPROVED** [review]
This is working now, but at your discretion, you may want to remove caching or add coverage reporting as discussed.

**Review by aherbst-broad — APPROVED** [review]

**Pull request merged**